src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+using Microsoft.JSInterop.WebAssembly;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Services
+{
+    public class LazyAssemblyLoader
+    {
+        internal const string GetDynamicAssemblies = "window.Blazor._internal.getLazyAssemblies";
+        internal const string ReadDynamicAssemblies = "window.Blazor._internal.readLazyAssemblies";
+
+        private List<string> _loadedAssemblyCache = new List<string>();
+
+        private readonly IServiceProvider _provider;
+
+        public LazyAssemblyLoader(IServiceProvider provider)
+        {
+            _provider = provider;
+        }
+
+        public async Task<IEnumerable<Assembly>> LoadAssembliesAsync(IEnumerable<string> assembliesToLoad)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Browser))
+            {
+                return await LoadAssembliesInClientAsync(assembliesToLoad);
+            }
+
+            return await LoadAssembliesInServerAsync(assembliesToLoad);
+        }
+
+        private Task<IEnumerable<Assembly>> LoadAssembliesInServerAsync(IEnumerable<string> assembliesToLoad)
+        {
+            var assemblies = AppDomain.CurrentDomain.GetAssemblies().Where(assembly =>
+                assembliesToLoad.Contains(assembly.GetName().Name + ".dll"));
```

Comment:
I'm not massively worried here because the total number of assemblies won't be huge, but this will re-iterate through `assembliesToLoad` for each existing assembly in the appdomain.

If you wanted to, you could build a hashset of out `assembliesToLoad` first, then use .Contains on that.

Reply:
Also it will evaluate this twice because `assemblies` is not a materialized collection. Again not a massive issue.